File: sdk/android/vira-android/src/main/kotlin/ViraExperience.kt
import android.content.Context
import android.view.View

/** Stable customer-facing Android entrypoint. Runtime/session/registry ownership stays inside Vira. */
class ViraExperience private constructor(
  val view: View,
  private val session: ViraAndroidRuntimeSession,
  private val surface: ViraAndroidSurfaceController,
) {
  fun dispose() {
    surface.dispose()
    session.dispose()
  }

  companion object {
    fun create(
      context: Context,
      mountEnvelopeJson: String,
      runtimeStateJson: String,
      permissionPolicyJson: String,
      host: ViraAndroidHostAdapter,
      renderers: List<ViraAndroidNativeRenderer>,
    ): Result<ViraExperience> = runCatching {
      val envelope = ViraAndroidMountEnvelope.decode(mountEnvelopeJson).getOrThrow()
      val runtimeState = ViraAndroidRuntimeCoreState.decode(runtimeStateJson).getOrThrow()
      val permissionPolicy = ViraAndroidPermissionPolicy.decode(permissionPolicyJson).getOrThrow()
      if (runtimeState.experienceId != envelope.document.id) {
        throw ViraAndroidIssue(
          ViraAndroidIssueCode.INVALID_RUNTIME_STATE,
          "$.runtimeState.experienceId",
          "runtime state belongs to a different Experience",
        )
      }
      val session = ViraAndroidRuntimeSession(envelope, host, runtimeState, permissionPolicy)
      try {
        val registry = ViraAndroidRendererRegistry.create(envelope, renderers).getOrThrow()
        val surface = ViraAndroidSurfaceController.create(context, session, registry).getOrThrow()
        ViraExperience(surface.view, session, surface)
      } catch (error: Throwable) {
        session.dispose()
        throw error
      }
    }
  }
}

File: sdk/ios/Sources/ViraIOS/ViraExperience.swift
#if canImport(UIKit)
import Foundation
import UIKit

/// Stable customer-facing iOS entrypoint. Runtime/session/registry ownership stays inside Vira.
@MainActor
public final class ViraExperience {
  public let view: UIView

  private let session: ViraIOSRuntimeSession
  private let surface: ViraIOSUIKitSurfaceController

  private init(session: ViraIOSRuntimeSession, surface: ViraIOSUIKitSurfaceController) {
    self.session = session
    self.surface = surface
    self.view = surface.view
  }

  public static func create(
    mountEnvelopeJSON: String,
    runtimeStateJSON: String,
    permissionPolicyJSON: String,
    host: ViraIOSHostAdapter,
    renderers: [any ViraIOSUIKitRenderer]
  ) -> Result<ViraExperience, ViraIOSIssue> {
    guard let envelopeData = mountEnvelopeJSON.data(using: .utf8),
          let stateData = runtimeStateJSON.data(using: .utf8),
          let permissionData = permissionPolicyJSON.data(using: .utf8) else {
      return .failure(.init(code: .invalidEnvelope, path: "$", message: "Vira Experience inputs must be UTF-8 JSON"))
    }

    let envelope: ViraIOSMountEnvelope
    switch ViraIOSMountEnvelope.decode(envelopeData) {
    case .failure(let issue): return .failure(issue)
    case .success(let value): envelope = value
    }
    let runtimeState: ViraIOSRuntimeCoreState
    switch ViraIOSRuntimeCoreState.decode(stateData) {
    case .failure(let issue): return .failure(issue)
    case .success(let value): runtimeState = value
    }
    guard runtimeState.experienceId == envelope.document.id else {
      return .failure(.init(
        code: .invalidRuntimeState,
        path: "$.runtimeState.experienceId",
        message: "runtime state belongs to a different Experience"
      ))
    }
    let permissionPolicy: ViraIOSPermissionPolicy
    switch ViraIOSPermissionPolicy.decode(permissionData) {
    case .failure(let issue): return .failure(issue)
    case .success(let value): permissionPolicy = value
    }

    let session: ViraIOSRuntimeSession
    do {
      session = try ViraIOSRuntimeSession(
        envelope: envelope,
        host: host,
        runtimeState: runtimeState,
        permissionPolicy: permissionPolicy
      )
    } catch let issue as ViraIOSIssue {
      return .failure(issue)
    } catch {
      return .failure(.init(code: .invalidEnvelope, path: "$", message: "Vira Experience session creation failed"))
    }

    switch ViraIOSUIKitSurfaceController.create(session: session, renderers: renderers) {
    case .failure(let issue):
      session.dispose()
      return .failure(issue)
    case .success(let surface):
      return .success(ViraExperience(session: session, surface: surface))
    }
  }

  public func dispose() {
    surface.dispose()
    session.dispose()
  }
}
#endif
